WordPress/gutenberg
# [TypeScript] More specific core-data selector types

## What?

This PR narrows down `core-data` selectors type signatures after somewhat general ones were introduced as a first step in https://github.com/WordPress/gutenberg/pull/41235.

## Why?

We want to have autocompletion and TypeScript support for core-data primitives as well as compile-time checks, see https://github.com/WordPress/gutenberg/pull/39025 for more context.

## Testing Instructions

* Confirm the CI checks are green
* Confirm the few runtime changes this PR introduces to make TS happy are sound

@noisysocks @sirreal @dmsnell @sarayourfriend
